django/db/backends/postgresql/base.py

```diff
@@ -266,23 +266,22 @@ def get_new_connection(self, conn_params):
                     f"Invalid transaction isolation level {isolation_level_value} "
                     f"specified. Use one of the psycopg.IsolationLevel values."
                 )
+        conn_params = {
+            "cursor_factory": ServerBindingCursor
+            if is_psycopg3 and options.get("server_side_binding")
+            else Cursor,
+            **conn_params,  # Last to allow the user to override cursor_factory.
+        }
```

Comment:
Can we use `setdefault()` instead of re-creating `conn_params`?
```suggestion
        conn_params.setdefault(
            "cursor_factory", 
            if is_psycopg3 and options.get("server_side_binding") is True
            else Cursor,
        }
```

We use `is True` intentionally, to avoid errors when passing wrong truthy values, e.g.
```
DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        # ...
        "OPTIONS": {
            "server_side_binding": "False",
        },
    },
}
```
which is quite common when folks use environment variables:
```
"server_side_binding": os.environ.get("USE_SSB")
```

Reply:
Using `setdefault` here would mutate the dictionary that’s owned by the caller.

Will put back `is True`.

Reply:
> Using `setdefault` here would mutate the dictionary that’s owned by the caller.

This should not be an issue when we will move it to the `get_connection_params()`, which is now a better place to handle this since we don't modify a connection directly but connection params.

Reply:
Perhaps it would make sense to do `conn_params.setdefault("cursor_factory", …)` within `get_connection_params` instead?

Reply:
> Perhaps it would make sense to do `conn_params.setdefault("cursor_factory", …)` within `get_connection_params` instead?

Since we create `conn_params` from scratch, we could do:
```diff
diff --git a/django/db/backends/postgresql/base.py b/django/db/backends/postgresql/base.py
index 3db3d6abd2..319dc63d05 100644
--- a/django/db/backends/postgresql/base.py
+++ b/django/db/backends/postgresql/base.py
@@ -208,7 +208,9 @@ class DatabaseWrapper(BaseDatabaseWrapper):
                     self.ops.max_name_length(),
                 )
             )
-        conn_params = {"client_encoding": "UTF8"}
+        conn_params = {"client_encoding": "UTF8", "cursor_factory": Cursor}
+        if is_psycopg3 and settings_dict["OPTIONS"].get("server_side_binding") is True:
+            conn_params["cursor_factory"] = ServerBindingCursor
         if settings_dict["NAME"]:
             conn_params = {
                 "dbname": settings_dict["NAME"],
```

Reply:
(Oops, yeah, working on bad train wifi.) Moved to `get_connection_params` with `setdefault`.

Reply:
> (Oops, yeah, working on bad train wifi.) Moved to `get_connection_params` with `setdefault`.

See https://github.com/django/django/pull/16736#discussion_r1160441232.

Reply:
Something’s wrong with that `conn_params = {"client_encoding": "UTF8"}` line. `conn_params` is unconditionally rewritten below, so that value is never used.

Reply:
> Something’s wrong with that `conn_params = {"client_encoding": "UTF8"}` line. `conn_params` is unconditionally rewritten below, so that value is never used.

Good catch :dart: Let's discuss it separately as it's not related with this issue.

Reply:
> > Something’s wrong with that `conn_params = {"client_encoding": "UTF8"}` line. `conn_params` is unconditionally rewritten below, so that value is never used.
> 
> Good catch dart Let's discuss it separately as it's not related with this issue.

It's a separate regression in 6a2165816394ab4bb259f6171e82417e098e97a6.

Reply:
https://github.com/django/django/pull/16362#discussion_r1160449022

Reply:
I created a ticket, see ticket-34470.